Use libwebauthn for JSON request parsing

AlfioEmanueleFresta opened the pull request:
This PR migrates JSON request parsing to use libwebauthn's `WebAuthnIDL::from_json()` trait instead of our custom parsing code. This removes ~700 lines of manual parsing in favour of the shared implementation.

## Changes

- Use `MakeCredentialRequest::from_json()` and `GetAssertionRequest::from_json()` from libwebauthn
- Remove intermediate parsing structs (`MakeCredentialOptions`, `GetCredentialOptions`, `CredentialDescriptor`, etc.)
- Pin libwebauthn to commit `d97c80d25bdb974472c40de5e5031db5946ad532` (from https://github.com/linux-credentials/libwebauthn/pull/155)

## Behavioral changes

### Default timeout
The default timeout when not specified by the relying party changes from **300s to 60s**:
- Old: https://github.com/linux-credentials/credentialsd/blob/03206d28a124fc6d36e52b946933ddb2db4810a7/credentialsd/src/dbus/model.rs#L211
- New: https://github.com/linux-credentials/libwebauthn/blob/d97c80d25bdb974472c40de5e5031db5946ad532/libwebauthn/src/ops/webauthn/timeout.rs#L3

### Allow list transports
Previously we cleared `transports` from credentials in the allow list as a workaround. This is no longer done - transports now pass through as-is. These are just UI hints and shouldn't affect functionality.

## Follow-up

- JSON *response* serialization will come in a follow-up PR 
- This PR depends on https://github.com/linux-credentials/libwebauthn/pull/155  
- Filed linux-credentials/libwebauthn#157 for a userVerification default issue

AlfioEmanueleFresta committed:
Use libwebauthn for JSON parsing

msirringhaus commented on credentialsd/src/credential_service/nfc.rs:
```diff
@@ -36,8 +36,11 @@ impl InProcessNfcHandler {
-            Ok(Some(hid_device)) => Ok(NfcStateInternal::Connected(hid_device)),
+            Ok(Some(nfc_device)) => Ok(NfcStateInternal::Connected(nfc_device)),
+            Ok(None) => {
+                let state = NfcStateInternal::Waiting;
```
I don't understand the change made here

msirringhaus commented on credentialsd/src/dbus/model.rs:
```diff
@@ -49,171 +60,48 @@ pub(super) fn create_credential_request_try_into_ctap2(
-        MakeCredentialRequest {
-            hash: client_data_hash,
-            origin,
+        String::from_utf8(make_cred_request.client_data_json()).map_err(|_| {
```
See my comment here https://github.com/linux-credentials/libwebauthn/pull/155#discussion_r2672162252 I think it would be sensible for `make_cred_request.client_data_json()` to return a String.

msirringhaus reviewed (dismissed):
Left some questions inline

AlfioEmanueleFresta committed:
Feedback: split domain and update libwebauthn

AlfioEmanueleFresta committed:
Cleaned up unused code

AlfioEmanueleFresta committed:
Update libwebauthn to 0.3.0

Switch from a git rev pin to the published 0.3.0 release on crates.io.

Adapt to new APIs:
- CableQrCodeDevice::new_transient now returns Result<Self, Error>.
- UvUpdate gained a PinNotSet variant; handle it (currently logs an
  error pending UI support for PIN setup).

AlfioEmanueleFresta committed:
Derive rpId from NavigationContext in gateway

AlfioEmanueleFresta commented:
@iinuwa this is now rebased on libwebauthn v0.3.0 and ready for review. If OK, I'll keep the libwebauthn API change (passing in the pre-parased Origin) as a follow up.

AlfioEmanueleFresta committed:
Apply cargo fmt

iinuwa commented on credentialsd/Cargo.lock:
Can you delete this file? This crate should share the lockfile with the workspace in the root directory.

iinuwa commented on credentialsd-common/Cargo.lock:
Let's delete this file too.

iinuwa reviewed (changes requested):
I think this seems fine, but I wasn't able to test due to https://github.com/linux-credentials/libwebauthn/issues/191. I do want to drop the extra lockfiles though, since the one at the workspace root is the effective one.

AlfioEmanueleFresta commented:
Thanks for your review @iinuwa.

I was able to reproduce your issue - it appears to be specific to SoloKeys Solo 2 firmware and JSON requests which include a `PublicKeyCredentialDescriptor` item (eg. within `allowlist`) with `transports` populated.

If it works with you, once https://github.com/linux-credentials/libwebauthn/pull/192 lands I'll publish a new libwebauthn version, then open a new credentialsd PR for the bump.

AlfioEmanueleFresta committed:
Remove unnecessary subdirectory Cargo.lock files

The workspace root Cargo.lock is the effective one.
